File: java/po/world_simulator/Map.java
package po.world_simulator;

import java.util.ArrayList;
import java.util.List;
import po.world_simulator.entities.Entity;

public class Map {

    private final Vector2[][] tileNeighbours;

    private final Entity[][] map;
    private final Vector2 size;

    public Map(Vector2 mapSize, boolean hexMode) {
        this.size = mapSize;
        if(hexMode){
            this.tileNeighbours = new Vector2[][]{
                {
                    new Vector2(-1, -1), new Vector2(0, -1),
                    new Vector2(-1, 0), new Vector2(1, 0),
                    new Vector2(-1, 1),  new Vector2(0, 1)
                },
                {
                    new Vector2(0, -1), new Vector2(1, -1),
                    new Vector2(-1, 0), new Vector2(1, 0),
                    new Vector2(0, 1),  new Vector2(1, 1)
                }
            };
        }
        else{
            this.tileNeighbours = new Vector2[][]{
                {
                    new Vector2(-1, -1), new Vector2(0, -1), new Vector2(1, -1),
                    new Vector2(-1, 0),                      new Vector2(1, 0),
                    new Vector2(-1, 1),  new Vector2(0, 1),  new Vector2(1, 1)
                },
                {
                    new Vector2(-1, -1), new Vector2(0, -1), new Vector2(1, -1),
                    new Vector2(-1, 0),                      new Vector2(1, 0),
                    new Vector2(-1, 1),  new Vector2(0, 1),  new Vector2(1, 1)
                }
            };
        }
        // W Javie używamy tablic 2D (lub list), inicjalizują się na null domyślnie
        this.map = new Entity[size.x][size.y];
    }

    public Entity getEntityAt(Vector2 position) {
        if (isTileOutOfBounds(position)) return null;
        return map[position.x][position.y];
    }

    public void placeEntityAt(Vector2 position, Entity entity) {
        if (!isTileOutOfBounds(position)) {
            map[position.x][position.y] = entity;
        }
    }

    public void removeEntityAt(Vector2 position) {
        if (!isTileOutOfBounds(position)) {
            map[position.x][position.y] = null;
        }
    }

    public boolean isTileOccupied(Vector2 position) {
        return !isTileOutOfBounds(position) && map[position.x][position.y] != null;
    }

    public boolean isTileOutOfBounds(Vector2 position) {
        return position.x < 0 || position.y < 0
            || position.x >= size.x || position.y >= size.y;
    }

    public Vector2 getPossibleMoveDirection(Vector2 position, boolean onlyFreeTiles) {
        List<Vector2> possibleMoves = getNeighbours(position, onlyFreeTiles, false);

        if (possibleMoves.isEmpty()) {
            return Vector2.ZERO;
        } else {
            Vector2 movePosition = possibleMoves.get((int) (Math.random() * possibleMoves.size()));
            return new Vector2(movePosition.x - position.x, movePosition.y - position.y);
        }
    }

    public List<Entity> getEntitiesAround(Vector2 position) {
        List<Vector2> possibleMoves = getNeighbours(position, false, true);
        List<Entity> entities = new ArrayList<>();

        for (Vector2 pos : possibleMoves) {
            entities.add(map[pos.x][pos.y]);
        }
        return entities;
    }

    public void move(Vector2 from, Vector2 to) {
        Entity temp = map[to.x][to.y];
        map[to.x][to.y] = map[from.x][from.y];
        map[from.x][from.y] = temp;
    }

    private List<Vector2> getNeighbours(Vector2 position, boolean onlyFreeTiles, boolean onlyOccupiedTiles) {
        List<Vector2> tiles = new ArrayList<>();

        for (Vector2 direction : tileNeighbours[position.y % 2]) {
            Vector2 neighbour = position.add(direction);
            if (isTileOutOfBounds(neighbour)) continue;

            Entity occupant = map[neighbour.x][neighbour.y];
            if (
                    (onlyFreeTiles && occupant == null)
                    || (!onlyFreeTiles && !onlyOccupiedTiles)
                    || (onlyOccupiedTiles && occupant != null)
            )
                tiles.add(neighbour);
        }
        return tiles;
    }
}

File: python/po/world_simulator/Map.py
import random
from .Vector2 import Vector2
import math

class Map:
    def __init__(self, mapSize, hexMode):
        self.size = mapSize
        self.hexMode = hexMode
        if hexMode:
            self.tileNeighbours = [
                [
                    Vector2(-1, -1), Vector2(0, -1),
                    Vector2(-1, 0), Vector2(1, 0),
                    Vector2(-1, 1),  Vector2(0, 1)
                ],
                [
                    Vector2(0, -1), Vector2(1, -1),
                    Vector2(-1, 0), Vector2(1, 0),
                    Vector2(0, 1),  Vector2(1, 1)
                ]
            ]
        else:
            self.tileNeighbours = [
                [
                    Vector2(-1, -1), Vector2(0, -1), Vector2(1, -1),
                    Vector2(-1, 0),                  Vector2(1, 0),
                    Vector2(-1, 1),  Vector2(0, 1),  Vector2(1, 1)
                ],
                [
                    Vector2(-1, -1), Vector2(0, -1), Vector2(1, -1),
                    Vector2(-1, 0),                  Vector2(1, 0),
                    Vector2(-1, 1),  Vector2(0, 1),  Vector2(1, 1)
                ]
            ]

        self.map = [[None for _ in range(self.size.y)] for _ in range(self.size.x)]


    def getEntityAt(self, position):
        if self.isTileOutOfBounds(position):
            return None
        return self.map[position.x][position.y]


    def placeEntityAt(self, position, entity):
        if not self.isTileOutOfBounds(position):
            self.map[position.x][position.y] = entity


    def removeEntityAt(self, position):
        if not self.isTileOutOfBounds(position):
            self.map[position.x][position.y] = None


    def isTileOccupied(self, position):
        return not self.isTileOutOfBounds(position) and self.map[position.x][position.y] is not None


    def isTileOutOfBounds(self, position):
        return position.x < 0 or position.y < 0 or position.x >= self.size.x or position.y >= self.size.y


    def getPossibleMoveDirection(self, position, onlyFreeTiles):
        possibleMoves = self.getNeighbours(position, onlyFreeTiles, False)

        if not possibleMoves:
            return Vector2.ZERO
        else:
            movePosition = possibleMoves[int(random.random() * len(possibleMoves))]
            return Vector2(movePosition.x - position.x, movePosition.y - position.y)


    def getEntitiesAround(self, position):
        possibleMoves = self.getNeighbours(position, False, True)
        entities = []

        for pos in possibleMoves:
            entities.append(self.map[pos.x][pos.y])
        return entities


    def move(self, from_pos, to_pos):
        temp = self.map[to_pos.x][to_pos.y]
        self.map[to_pos.x][to_pos.y] = self.map[from_pos.x][from_pos.y]
        self.map[from_pos.x][from_pos.y] = temp


    def getNeighbours(self, position, onlyFreeTiles, onlyOccupiedTiles):
        tiles = []

        for direction in self.tileNeighbours[position.y % 2]:
            neighbour = position.add(direction)
            
            if self.isTileOutOfBounds(neighbour):
                continue

            occupant = self.map[neighbour.x][neighbour.y]
            if (
                (onlyFreeTiles and occupant is None)
                or (not onlyFreeTiles and not onlyOccupiedTiles)
                or (onlyOccupiedTiles and occupant is not None)
            ):
                tiles.append(neighbour)
                
        return tiles

    # calculate total moves needed to get from from_pos to to_pos, assuming no obstacles
    def distanceTo(self, from_pos, to_pos):
        if not self.hexMode:
            return max(abs(from_pos.x - to_pos.x), abs(from_pos.y - to_pos.y))
        else:
            dx = from_pos.x - to_pos.x
            dy = abs(from_pos.y - to_pos.y)
            if from_pos.y % 2 == 0:
                return dy + max(0, abs(dx) - (dy + 1 * math.copysign(1, dx)) // 2)
            else:
                return dy + max(0, abs(dx) - dy // 2)


    def getMoveDirectionTowards(self, from_pos, to_pos):
        if not self.hexMode:
            dx = to_pos.x - from_pos.x
            dy = to_pos.y - from_pos.y
            move_x = 0 if dx == 0 else (1 if dx > 0 else -1)
            move_y = 0 if dy == 0 else (1 if dy > 0 else -1)
            return Vector2(move_x, move_y)
        else:
            # Hex grid movement logic
            best_direction = Vector2.ZERO
            min_distance = math.inf

            for direction in self.tileNeighbours[from_pos.y % 2]:
                new_position = from_pos.add(direction)
                distance = self.distanceTo(new_position, to_pos)
                if distance < min_distance and not self.isTileOutOfBounds(new_position):
                    min_distance = distance
                    best_direction = direction

            return best_direction